internal/installercache/installercache.go

```diff
@@ -80,22 +111,87 @@ func (rl *Release) Cleanup(ctx context.Context) {
+		return nil, errors.New("config.InstallerCacheEvictionThreshold must not be zero")
+	}
+	if config.MaxCapacity > 0 && config.MaxReleaseSize == 0 {
+		return nil, fmt.Errorf("config.MaxReleaseSize (%d bytes) must not be zero", config.MaxReleaseSize)
+	}
+	if config.MaxCapacity > 0 && config.MaxReleaseSize > config.MaxCapacity {
+		return nil, fmt.Errorf("config.MaxReleaseSize (%d bytes) must not be greater than config.MaxCapacity (%d bytes)", config.MaxReleaseSize, config.MaxCapacity)
+	}
 	return &Installers{
 		log:             log,
-		storageCapacity: storageCapacity,
-		cacheDir:        cacheDir,
 		eventsHandler:   eventsHandler,
-	}
+		diskStatsHelper: diskStatsHelper,
+		config:          config,
+	}, nil
 }
 
 // Get returns the path to an openshift-baremetal-install binary extracted from
 // the referenced release image. Tries the mirror release image first if it's set. It is safe for concurrent use. A cache of
 // binaries is maintained to reduce re-downloading of the same release.
 func (i *Installers) Get(ctx context.Context, releaseID, releaseIDMirror, pullSecret string, ocRelease oc.Release, ocpVersion string, clusterID strfmt.UUID) (*Release, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+			release, err := i.get(releaseID, releaseIDMirror, pullSecret, ocRelease, ocpVersion, clusterID)
+			if err == nil {
+				return release, nil
+			}
+			_, isCapacityError := err.(*errorInsufficientCacheCapacity)
+			if !isCapacityError {
+				return nil, errors.Wrapf(err, "failed to get installer path for release %s", releaseID)
+			}
+			time.Sleep(i.config.ReleaseFetchRetryInterval)
+		}
+	}
+}
+
+func (i *Installers) getDiskUsageIncludingHardlinks() (uint64, error) {
+	usedBytes, _, err := i.diskStatsHelper.GetDiskUsage(i.config.CacheDir)
+	if err != nil && !os.IsNotExist(err) {
+		return 0, errors.Wrapf(err, "could not determine disk usage information for cache dir %s", i.config.CacheDir)
+	}
+	return usedBytes, nil
+}
+
+func (i *Installers) extractReleaseIfNeeded(path, releaseID, releaseIDMirror, pullSecret, ocpVersion string, ocRelease oc.Release, release *Release) error {
+	_, err := os.Stat(path)
+	if err == nil {
+		return nil // No error indicates that the release already exists
+	}
+	if !os.IsNotExist(err) {
+		return err
+	}
+	usedBytes, err := i.getDiskUsageIncludingHardlinks()
+	if err != nil {
+		return errors.Wrapf(err, "could not determine disk usage information for cache dir %s", i.config.CacheDir)
+	}
+	if i.shouldEvict(int64(usedBytes)) && !i.evict() {
+		return &errorInsufficientCacheCapacity{Message: fmt.Sprintf("insufficient capacity in %s to store release", i.config.CacheDir)}
+	}
+	extractStartTime := time.Now()
+	_, err = ocRelease.Extract(i.log, releaseID, releaseIDMirror, i.config.CacheDir, pullSecret, ocpVersion)
+	if err != nil {
+		return err
+	}
+	release.extractDuration = time.Since(extractStartTime).Seconds()
+	release.cached = false
```

Comment:
Generally I prefer returning extra variables over side effects like setting fields on pointer arguments.
This is definitely a personal preference thing, but what do you think about returning variables for `cached` and `extractDuration` where `err == nil` and `cached == false` would mean `extractDuration` should be valid?

Reply:
I'll switch it up so we avoid side effect and use return values, I can see that value in this